### PR TITLE
Move `Shell::Add/RemoveView` to `PlatformView` and refine embedder API doc

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR moves the methods to add or remove views from `Shell` to `PlatformView`. By design, the `Shell` is supposed to be a messenger that glues classes together, while `PlatformView` is the operator that embedders that do not use the embedder API should operate on. The current design was made due to lack of knowledge to this design.

This also makes `PlatformView` aware of views, which might be a prerequisite to https://github.com/flutter/engine/pull/51925.

This PR also adds some details to embedder API `AddView` and `RemoveView`.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dkwingsmt** [comment]
@chinmaygarde Can you take a look? I'm not sure whether this is a necessary change. I'm proposing this only because I supposed the `Shell` should have as few methods as possible and the add/removeView methods should have existed on `PlatformView` to be consistent with other existing methods, such as `DispatchPointerDataPacket`.

File: shell/common/platform_view.h [hunk 1]
```diff
@@ -51,13 +51,17 @@ namespace flutter {
...
+  using AddViewCallback = std::function<void(bool added)>;
+  using RemoveViewCallback = std::function<void(bool removed)>;
   //----------------------------------------------------------------------------
   /// @brief      Used to forward events from the platform view to interested
   ///             subsystems. This forwarding is done by the shell which sets
   ///             itself up as the delegate of the platform view.
   ///
   class Delegate {
    public:
+    using AddViewCallback = PlatformView::AddViewCallback;
```
Comment (chinmaygarde) [inline shell/common/platform_view.h:63]: Nit: Is this necessary? You can just use the typedef from the parent.

File: shell/common/platform_view.h [hunk 2]
```diff
@@ -51,13 +51,17 @@ namespace flutter {
 ///
 class PlatformView {
  public:
+  using AddViewCallback = std::function<void(bool added)>;
```
Comment (chinmaygarde) [inline shell/common/platform_view.h:54]: What do the `added` and `removed` arguments to the callback mean? If `added` is false, why invoke the `AddViewCallback`? Perhaps document that?

File: shell/common/platform_view.h (hunk 1 above)
Comment (dkwingsmt) [reply shell/common/platform_view.h:63]: The compiler requires `Shell` to use `PlatformView::AddViewCallback` and does not accept simply `AddViewCallback` without it. It's definitely not necessary, and I'm down to remove it.

File: shell/common/platform_view.h (hunk 2 above)
Comment (dkwingsmt) [reply shell/common/platform_view.h:54]: Both the add and the remove callbacks rely on the result. For example, in the Windows embedder as implemented in https://github.com/flutter/engine/pull/51923
* The embedder blocks on the `AddView` callback until a result is reported, which decides whether the returned `FlutterWindowsView` is null.
* The embedder also pre-allocate resources after calling `AddView` before the callback, and if the result turns out a failure, then the resources are deallocated. The pre-allocation ensures that the framework can start operating the view as soon as they're aware of it.

File: shell/platform/embedder/embedder.h [hunk 3]
```diff
@@ -2601,9 +2615,18 @@ FlutterEngineResult FlutterEngineAddView(FLUTTER_API_SYMBOL(FlutterEngine)
...
+///             |info.remove_view_callback| with whether the operation is
+///             successful.
+///
+///             This operation is asynchronous. The embedder should not
+///             deallocate resources until the |callback| is invoked.
+///
+///             Do not use for implicit views, which are never removed
+///             throughout the lifetime of the app.  Removing
+///             |kFlutterImplicitViewId| or an non-existent view ID will fail,
+///             indicated by |callback| returning false.
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2629]:
```suggestion
///             indicated by |info.remove_view_callback| returning false.
```

File: shell/platform/embedder/embedder.h [hunk 4]
```diff
@@ -2601,9 +2615,18 @@ FlutterEngineResult FlutterEngineAddView(FLUTTER_API_SYMBOL(FlutterEngine)
...
+///             inform Dart about the removal. Finally, it invokes
+///             |info.remove_view_callback| with whether the operation is
+///             successful.
+///
+///             This operation is asynchronous. The embedder should not
+///             deallocate resources until the |callback| is invoked.
+///
+///             Do not use for implicit views, which are never removed
+///             throughout the lifetime of the app.  Removing
+///             |kFlutterImplicitViewId| or an non-existent view ID will fail,
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2628]: Folks copy this file into their own projects, they won't have this `kFlutterImplicitViewId` constant. I would use a little more generic language here:

```suggestion
///             the implicit view ID or an non-existent view ID will fail,
```

File: shell/platform/embedder/embedder.h [hunk 5]
```diff
@@ -2601,9 +2615,18 @@ FlutterEngineResult FlutterEngineAddView(FLUTTER_API_SYMBOL(FlutterEngine)
...
-///             This is an asynchronous operation. The view's resources must not
-///             be cleaned up until the |remove_view_callback| is invoked with
-///             a |removed| value of `true`.
+///             This operation notifies the engine to deallocate resources and
+///             inform Dart about the removal. Finally, it invokes
+///             |info.remove_view_callback| with whether the operation is
+///             successful.
+///
+///             This operation is asynchronous. The embedder should not
+///             deallocate resources until the |callback| is invoked.
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2624]:
```suggestion
///             deallocate resources until the |info.remove_view_callback| is invoked.
```

Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2624]: The previous wording was more precise as to when the embedder can destroy a view's resources:

> The view's resources must not be cleaned up until the |remove_view_callback| is invoked with a |removed| value of `true`

This is important as the engine can continue presenting to the view if `FlutterEngineRemoveView` returned something other than `kSuccess`. Consider:

```suggestion
///             deallocate resources until the |info.remove_view_callback| is invoked with a |removed| value of `true`.
```

File: shell/platform/embedder/embedder.h [hunk 6]
```diff
@@ -2601,9 +2615,18 @@ FlutterEngineResult FlutterEngineAddView(FLUTTER_API_SYMBOL(FlutterEngine)
 //------------------------------------------------------------------------------
 /// @brief      Removes a view.
 ///
-///             This is an asynchronous operation. The view's resources must not
-///             be cleaned up until the |remove_view_callback| is invoked with
-///             a |removed| value of `true`.
+///             This operation notifies the engine to deallocate resources and
+///             inform Dart about the removal. Finally, it invokes
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2619]: I would remove this bit:

> This operation notifies the engine to deallocate resources and inform Dart about the removal.

In my opinion those are details unimportant to an embedder that are already implied by "Removes a view" above.

File: shell/platform/embedder/embedder.h [hunk 7]
```diff
@@ -2582,9 +2582,23 @@ FlutterEngineResult FlutterEngineRunInitialized(
 //------------------------------------------------------------------------------
 /// @brief      Adds a view.
 ///
-///             This is an asynchronous operation. The view should not be used
-///             until the |add_view_callback| is invoked with an `added` of
-///             `true`.
+///             This operation notifies the engine to allocate resources and
+///             inform Dart about the view, and on success, schedules a new
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2586]: I would remove this bit:

> This operation notifies the engine to allocate resources and inform Dart about the view

In my opinion those are details unimportant to an embedder and are already implied by "Adds a view" above.

File: shell/platform/embedder/embedder.h [hunk 8]
```diff
@@ -2582,9 +2582,23 @@ FlutterEngineResult FlutterEngineRunInitialized(
...
+///             frame. Finally, it invokes |info.add_view_callback| with whether
+///             the operation is successful.
+///
+///             This operation is asynchronous; avoid using the view until
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2590]: I would use a different word than "avoid". The Dart app's `PlatformDispatcher.instance.views` won't have this view until the callback is invoked with an `added` of `true`. The Dart app is _unable_ to use the view until then.

File: shell/platform/embedder/embedder.h [hunk 9]
```diff
@@ -2582,9 +2582,23 @@ FlutterEngineResult FlutterEngineRunInitialized(
...
+///             |info.add_view_callback| returns true. The embedder should
+///             prepare resources in advance but be ready to clean up on
+///             failure.
+///
+///             The callback is called on a different thread managed by the
+///             engine. The embedder must re-thread if needed.
+///
+///             Do not use for implicit views, which are added internally during
+///             initialization. Adding the implicit view ID or an existing view
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2599]: Attempting to add the implicit view ID will return a `kInvalidArguments` immediately. The callback will not be invoked.

File: shell/platform/embedder/embedder.h [hunk 10]
```diff
@@ -2582,9 +2582,23 @@ FlutterEngineResult FlutterEngineRunInitialized(
...
+///             ID will fail, indicated by |info.add_view_callback| returning
+///             false.
```
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2601]: What do you think of this copy?

>  This is an asynchronous operation. The view should not be used
> until the |info.add_view_callback| is invoked with an `added` of `true`.
>
> A frame is scheduled if the operation succeeds.
>
> The callback is invoked on a thread managed by the engine. The embedder should re-thread if needed.
>
> Attempting to add a view with an already existing view ID will fail,
> and |info.add_view_callback| will be invoked with an |added| of `false`.

File: shell/platform/embedder/embedder.h (hunk 3 above)
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:2629]: What do you think of this copy?

> This is an asynchronous operation. The view's resources must not
> be cleaned up until |info.remove_view_callback| is invoked with
> a |removed| value of `true`.
>
> The callback is invoked on a thread managed by the engine.
> The embedder should re-thread if needed.
>
> Attempting to remove a view with a non-existent view ID will fail,
and |info.remove_view_callback| will be invoked with a |removed| value of `false`.

**Comment by dkwingsmt** [comment]
I kept the statement "The embedder should prepare resources in advance but be ready to clean up on failure." since I think this is a very important implementation suggestion.

File: shell/platform/embedder/embedder.h (hunk 8 above)
Comment (dkwingsmt) [reply shell/platform/embedder/embedder.h]: Although the Dart app is unable to use the view until then, the embedder can use it by sending events. (However your version has sentences with similar meaning so it's fine.)

**Review by loic-sharma — APPROVED** [review]
LGTM, thanks for cleaning this up!

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**